---
title: "Bulk RNA-seq processing"
author: "Yann Vanrobaeys"
date: "2024-08-27"
output: 
  html_document:
    toc: true
    toc_depth: 3
  html_notebook: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(tidy=FALSE, cache=TRUE,
                      dev="png", message=FALSE, error=FALSE, warning=TRUE)
```

# R processing of gene expression data

## Load libraries

```{r, warning=FALSE,error=FALSE,message=FALSE}
library(openxlsx)
library(tidyverse)
library(reprex)
library(matrixStats)
library(XML)
library(ggrepel)
library(rtracklayer)
library(DESeq2)
library(apeglm)
library(ComplexHeatmap)
library(tximport)
library(edgeR)
library(gprofiler2)
library(cluster)
library(stringr)
library(fgsea)
library(UpSetR)
```

# load IGB libraries

Loading Custom code

data_formatting_tools.R: functions for organizing, combining, and writing data tables
analysis_tools.R: functions for running differential expression and PCA
plots.R: functions for generating figures

setting the scipen option as such makes R display numbers instead of shortening to scientific notation ("0.01" instead of "1e-02")

```{r}
source('Rcode/data_formatting_tools.R')
source('Rcode/analysis_tools.R')
source('Rcode/plots.R')
source('Rcode/common.R')
source('Rcode/ssGSEAProjection.Library.R')

options(scipen = 9)
```

## Prepare Annotation - gencode primary assembly basic annotation

If this was done previously it does not need to be repeated if files are available.
Comment out the block of code for either Human or Mouse.

### Human genome annotation

```{r}
# # Import the GTF file for the human genome
# z <- import("https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_43/gencode.v43.primary_assembly.basic.annotation.gtf.gz")
# 
# # Create tx2gene dataframe
# tx2gene <- as.data.frame(z) %>% select(c('transcript_id','gene_id'))
# tx2gene <- unique(tx2gene)
# tx2gene <- tx2gene %>% filter(!is.na(transcript_id))
# 
# # Create gene2typesym dataframe
# gene2typesym <- as.data.frame(z) %>% select(c('gene_id','gene_type','gene_name'))
# gene2typesym <- unique(gene2typesym)
# 
# # Create gene2chrPos dataframe
# gene2chrPos <- as.data.frame(z) %>% select(c('gene_id','seqnames','start','end'))
# gene2chrPos <- unique(gene2chrPos)
# 
# # Save to Excel files
# write.xlsx(tx2gene,colNames=TRUE, quote=FALSE, rowNames=FALSE, file="hg38_gencode43_tx2gene.xlsx")
# write.xlsx(gene2typesym,colNames=TRUE, quote=FALSE, rowNames=FALSE, file="hg38_gencode43_gene2typesym.xlsx")
# write.xlsx(gene2chrPos,colNames=TRUE, quote=FALSE, rowNames=FALSE, file="hg38_gencode43_gene2chrPos.xlsx")
```

### Mouse genome annotation

```{r}
# # Import the GTF file for the mouse genome
# z <- import("https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M33/gencode.vM33.primary_assembly.basic.annotation.gtf.gz")
# 
# # Create tx2gene dataframe
# tx2gene <- as.data.frame(z) %>% select(c('transcript_id','gene_id'))
# tx2gene <- unique(tx2gene)
# tx2gene <- tx2gene %>% filter(!is.na(transcript_id))
# 
# # Create gene2typesym dataframe
# gene2typesym <- as.data.frame(z) %>% select(c('gene_id','gene_type','gene_name'))
# gene2typesym <- unique(gene2typesym)
# 
# # Create gene2chrPos dataframe
# gene2chrPos <- as.data.frame(z) %>% select(c('gene_id','seqnames','start','end'))
# gene2chrPos <- unique(gene2chrPos)
# 
# # Save to Excel files
# write.xlsx(tx2gene, colNames=TRUE, quote=FALSE, rowNames=FALSE, file="mm10_gencodeM33_tx2gene.xlsx")
# write.xlsx(gene2typesym, colNames=TRUE, quote=FALSE, rowNames=FALSE, file="mm10_gencodeM33_gene2typesym.xlsx")
# write.xlsx(gene2chrPos, colNames=TRUE, quote=FALSE, rowNames=FALSE, file="mm10_gencodeM33_gene2chrPos.xlsx")
```

# Importing RNA-Seq data

## Read in star_salmon quantitation

Creating and loading samples.xlsx metadata file based on experiment.

```{r}
dir <- getwd()
list.files(dir)

samples <- read.xlsx("ExampleMetadata_samples.xlsx")
samples

files <- file.path(dir,samples$Folder, "quant.sf")
files

names(files) <- paste0(samples$Sample)
all(file.exists(files))

txi <- tximport(files, type = "salmon", tx2gene = tx2gene)
names(txi)

intCt <- round((txi$counts),0)
colnames(intCt)<-paste0(colnames(intCt),".intCt")

l2tpm <- log((txi$abundance+1),2)
colnames(l2tpm)<-paste0(colnames(l2tpm),".l2tpm")
dumpDat <- merge(intCt, l2tpm, by=0, all=TRUE)

# #For creating a matrix of CPMs within edgeR, the following code chunk can be used:
# y <- DGEList(txi$counts)
# cpms <- cpm(y)
# l2cpm <- log((cpms+1),2)
# colnames(l2cpm)<-paste0(colnames(l2cpm),".l2cpm")
# 
# dumpDat <- merge(intCt, l2cpm, by=0, all=TRUE)
```

## Boxplot of all log2 tpm data to check quality

This plot includes all transcripts/genes in the annotation target.

```{r, fig.width=6,fig.height=6}
l2tpm <- select(dumpDat, ends_with(".l2tpm"))

ggplot(stack(as.data.frame(l2tpm)), aes(x = ind, y = values)) + geom_boxplot() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ggtitle("l2tpm boxplot") 
```

## Annotate assembled data

Once quantified, the gene level salmon data needs to be associated with additional information to facilitate interpretation
Note: The human annotation file may contain a few examples of multiple gene/type annotations associated with the same EnsID.
This can duplicate rows in dumpData during merge, check dimensions of objects.

```{r}
assembleDat <- merge(x=dumpDat,y=gene2typesym, by.x=c("Row.names"), by.y=c("gene_id"), all.x=TRUE, all.y = FALSE)
assembleDat <- mutate(assembleDat, Avg = rowMeans(select(assembleDat, ends_with(".l2tpm")), na.rm = TRUE))
assembleDat <- mutate(assembleDat, Var = rowVars(as.matrix(select(assembleDat, ends_with(".l2tpm")), na.rm = TRUE)))
assembleDat <- assembleDat %>% mutate(LowExpLowVar = ifelse(Avg <= 0.1 & Var <= 0.1,"Yes","No"))
```

## Adding MPG column

In some cases, multiple geneids will map to the same gene symbol. In these cases, it is helpful to select a representative geneID for each group of gene symbols. The code block below iterates through each unique gene symbol, compares all of the geneids associated with the same gene symbol and identifies the geneid with the highest average expression. This gene is then set as the representative (max-per-gene) geneid for that symbol. Note: If multiple geneids have the same highest avg expression, the first gene that appears in the dataframe is taken

The code below performs the following steps:
  1) Create a vector called "mpg" of "No" values the same length as the number of rows in the dataframe assembleData
  2) Loop over the assembleDat object ... identifying index values of duplicate genes, updating mpg vector
  3) Merge mpg vector to assemblDat object with setting column name as "MPG"


```{r}
mpg <- rep('No', nrow(assembleDat))

for (i in unique(assembleDat[,"gene_name"])){
    if (is.na(i)){next}
    idxs <- which(assembleDat[,"gene_name"] == i)
    max_idx <- which(assembleDat$Avg[idxs] == max(assembleDat$Avg[idxs]))[1]

    mpg[idxs[max_idx]] <- 'Yes'
}

assembleDat$MPG <- mpg
```

## Additional boxplots after data filtering

This plot uses the annotation and expression level calculations to select only expressed protein coding genes

```{r, fig.width=6,fig.height=6}
l2tpm.select <- assembleDat %>% filter(MPG == "Yes" &
                                   gene_type == "protein_coding" &
                                   LowExpLowVar == "No") %>% select(ends_with(".l2tpm"))

#l2tpm.select <- assembleDat %>% filter(gene_type == "protein_coding") %>% select(ends_with(".l2tpm"))

dim(l2tpm.select)

ggplot(stack(as.data.frame(l2tpm.select)), aes(x = ind, y = values)) + geom_boxplot() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + ggtitle("l2tpm Expressed Protein coding boxplot") 
```

# Differential Expression Testing

## Create dds object from tximport data using Condition to define samples

```{r}
dds <- DESeqDataSetFromTximport(txi = txi,
                                     colData=samples,
                                     design = ~ Infection
                                     )

dds$Infection <- relevel(dds$Infection, "Mock")
```

### Prepare an exploratory PCA plot

By default, plotPCA using the 500 most variable genes for PCA 1 vs PCA 2 plots
this can be adjusted with ntop argument

```{r}
vsd <- vst(dds, blind=FALSE)
 
samples$Day <- as.factor(samples$Day)
 
plotPCA(vsd, intgroup="Condition", ntop=500) + geom_point(size = 4) + ggtitle("Condition")
plotPCA(vsd, intgroup="Infection", ntop=500) + geom_point(size = 4) + ggtitle("Infection")
plotPCA(vsd, intgroup="Run", ntop=500) + geom_point(size = 4) + ggtitle("Run")
plotPCA(vsd, intgroup="MappingRate", ntop=500) + geom_point(size = 4) + ggtitle("Mapping Rate")
plotPCA(vsd, intgroup="ReadCount", ntop=500) + geom_point(size = 4) + ggtitle("Read Count")

pcaData <- plotPCA(vsd, intgroup="Run", returnData=TRUE)
write.xlsx(pcaData, colNames=TRUE, rowNames=TRUE, overwrite = TRUE, file="pca.xlsx")

pdf('PCA.pdf',width=10, height=10)
PCAfig
dev.off()
```

### PCA plots with sample labels

In the example below, geom_label_repel is used to label points and axes are adjusted with the scale_x and scale_y components
adjust values according to requirements presented by the data

```{r,fig.height=12,fig.width=14}
plotPCA(vsd, intgroup="Condition", ntop=500) + 
  geom_point(size = 4) +
  geom_label_repel(aes(label = Condition), vjust = -0.5, hjust = 0.5, 
                   force = 2, min.segment.length = 0.1, max.overlaps = 100) +
  scale_y_continuous(limits = c(-34, 34), breaks = seq(-34, 34, by = 2)) +
scale_x_continuous(limits = c(-64, 64), breaks = seq(-64, 64, by = 2)) +
  ggtitle("Condition")
```


### Manual calculation of PCA results from Variance Stabilizing Transformed (VSD) data

```{r}
# We begin by specifying the number of top variable genes (`ntop`) to use in the PCA analysis.
ntop <- 500

# Extract the variance-stabilizing transformation (VSD) matrix from the `vsd` object.
mat <- assay(vsd)

# Calculate row variances (gene-wise variances) from the matrix.
mat.rv <- rowVars(mat)

# Select the top `ntop` most variable genes by ordering row variances in decreasing order.
rv.sel <- order(mat.rv, decreasing = TRUE)[seq_len(min(ntop, length(mat.rv)))]

# Extract the matrix corresponding to the top variable genes.
mat.sel <- assay(vsd)[rv.sel,]

# Perform Principal Component Analysis (PCA) on the transposed matrix (genes in columns, samples in rows).
pca.mat <- prcomp(t(mat.sel))

# Convert the PCA results (principal components for each sample) to a data frame.
pp.manual <- as.data.frame(pca.mat$x)

# Print the PCA results for inspection.
pp.manual

# Save the PCA results (sample principal components) to an Excel file.
write.xlsx(pp.manual, colNames = TRUE, rowNames = TRUE, overwrite = TRUE, file = "pca.manual.xlsx")

# Save the PCA loadings (the contribution of each gene to the principal components) to a separate Excel file.
write.xlsx(as.data.frame(pca.mat$rotation), colNames = TRUE, rowNames = TRUE, overwrite = TRUE, file = "pca.manual.rot.xlsx")
```

### Manual calculation of PCA results from log2 TPM-normalized data

```{r}
# Filter the dataset `assembleDat` to include only protein-coding genes with sufficient expression and variance.
# Keep the log2-transformed TPM values (columns ending with ".l2tpm").
l2tpm.pca <- assembleDat %>% filter(MPG == "Yes" &
                                   gene_biotype == "protein_coding" &
                                   LowExpLowVar == "No") %>% select(Row.names, ends_with(".l2tpm"))

# Set the gene names (row names) as row names for the `l2tpm.pca` matrix.
l2tpm.pca <- column_to_rownames(l2tpm.pca, "Row.names")

# Convert the data frame to a matrix for further analysis.
l2tpm.pca <- as.matrix(l2tpm.pca)

# Define the number of top variable genes (`ntop`) to use in this PCA calculation.
ntop <- 1500

# Calculate the row variances (gene-wise variances) for the log2 TPM matrix.
l2tpm.rv <- rowVars(l2tpm.pca)

# Select the top `ntop` most variable genes based on the variances.
l2tpm.sel <- order(l2tpm.rv, decreasing = TRUE)[seq_len(min(ntop, length(l2tpm.rv)))]

# Subset the VSD matrix for the top variable genes.
l2tpm.pca <- assay(vsd)[l2tpm.sel,]

# Perform PCA on the transposed matrix (genes in columns, samples in rows).
pca.l2tpm <- prcomp(t(l2tpm.pca))

# Convert the PCA results (principal components for each sample) to a data frame.
pp.l2 <- as.data.frame(pca.l2tpm$x)

# Print the PCA results for inspection.
pp.l2
```


### Run tests vs Mock

```{r}
dds <- DESeq(dds)
resultsNames(dds)
```

### Virus vs Mock

```{r}
vs.Mock.comparisons <- resultsNames(dds)[c(2,3,4)]
vs.Mock.comparisons
```
### Assemble results for vs Mock

```{r}
for (i in vs.Mock.comparisons){
  res <- as.data.frame(lfcShrink(dds, coef=i, type="apeglm"))

  compar_name <- i
  cleanCompar <- str_replace(compar_name, "Infection_", "std_")
  stat_col <- paste('WaldStatistic', compar_name, sep = '_')
  stat <- dds@rowRanges@elementMetadata[,stat_col]
  res$stat <- stat

  res <- res %>% dplyr::select(c("log2FoldChange", "pvalue", "padj","stat"))
  res <- res %>% dplyr::mutate(absFC = abs(log2FoldChange))
  res <- res %>% dplyr::mutate(negLogP = -log10(padj))

  colnames(res) <- paste(cleanCompar, colnames(res), sep = '.')
  res <- tibble::rownames_to_column(res, "RowName")
  assembleDat <- merge(x=assembleDat,y=res, by.x=c("Row.names"),by.y=c("RowName"), all=TRUE)
}
```

## Re-level to PR

```{r}
dds$Infection <- relevel(dds$Infection, "PR")
```

### Run tests vs PR

```{r}
dds <- DESeq(dds)
resultsNames(dds)
```

### Virus vs PR

```{r}
vs.PR.comparisons <- resultsNames(dds)[c(3,4)]
vs.PR.comparisons
```

### Assemble results for vs PR

```{r}
for (i in vs.PR.comparisons){
  res <- as.data.frame(lfcShrink(dds, coef=i, type="apeglm"))

  compar_name <- i
  cleanCompar <- str_replace(compar_name, "Infection_", "std_")
  stat_col <- paste('WaldStatistic', compar_name, sep = '_')
  stat <- dds@rowRanges@elementMetadata[,stat_col]
  res$stat <- stat

  res <- res %>% dplyr::select(c("log2FoldChange", "pvalue", "padj","stat"))
  res <- res %>% dplyr::mutate(absFC = abs(log2FoldChange))
  res <- res %>% dplyr::mutate(negLogP = -log10(padj))

  colnames(res) <- paste(cleanCompar, colnames(res), sep = '.')
  res <- tibble::rownames_to_column(res, "RowName")
  assembleDat <- merge(x=assembleDat,y=res, by.x=c("Row.names"),by.y=c("RowName"), all=TRUE)
}
```

## Re-level to UG

```{r}
dds$Infection <- relevel(dds$Infection, "UG")
```

### Run tests vs UG

```{r}
dds <- DESeq(dds)
resultsNames(dds)
```

### Virus vs UG

```{r}
vs.UG.comparisons <- resultsNames(dds)[c(4)]
vs.UG.comparisons
```

### Assemble results for vs UG

```{r}
for (i in vs.UG.comparisons){
  res <- as.data.frame(lfcShrink(dds, coef=i, type="apeglm"))

  compar_name <- i
  cleanCompar <- str_replace(compar_name, "Infection_", "std_")
  stat_col <- paste('WaldStatistic', compar_name, sep = '_')
  stat <- dds@rowRanges@elementMetadata[,stat_col]
  res$stat <- stat

  res <- res %>% dplyr::select(c("log2FoldChange", "pvalue", "padj","stat"))
  res <- res %>% dplyr::mutate(absFC = abs(log2FoldChange))
  res <- res %>% dplyr::mutate(negLogP = -log10(padj))

  colnames(res) <- paste(cleanCompar, colnames(res), sep = '.')
  res <- tibble::rownames_to_column(res, "RowName")
  assembleDat <- merge(x=assembleDat,y=res, by.x=c("Row.names"),by.y=c("RowName"), all=TRUE)
}
```


# Differential Expression Testing - Complex model

If PCA reveals batch effect, include it in the complex dds design

```{r}
dds_run <- DESeqDataSetFromMatrix(countData = intCt,
                              colData=samples,
                              design = ~ Run + Infection)

dds_run$Group <- relevel(dds_run$Group, "Mock.7")
```

Repeat previous steps testing, re-leveling, etc...

## Write differential expression results to excel

```{r}
assembleDat <- assembleDat %>% mutate(ENSid = str_extract(Row.names, "^[^.]+"))
write.xlsx(assembleDat,colNames=TRUE, quote=FALSE, rowNames=FALSE, file="assemblDat.xlsx")
```

# Volcano plots

You can also use the shiny app wit hthe tab "Volcano Plot" and upload your own differential gene expression table to produce a volcano plot: 

https://yannvrb.shinyapps.io/BioMicro_data_visualization/

## Prepare gene expression data for plotting

```{r}
vol_df <- assembleDat %>% dplyr::filter(MPG == "Yes" & 
                                    gene_type == "protein_coding" & 
                                    LowExpLowVar == "No") %>% dplyr::select(gene_name, ends_with(".log2FoldChange"),ends_with(".padj"))
```

## Adjust the data table for better plotting

### Replaces NA value padj columns with 1 and NA value log2FoldChange with 0 and sets y axis according to max(abs(LFC))

```{r}
vol_df.edit <- vol_df

vol_df.edit <- vol_df.edit <- vol_df.edit %>% mutate_at(vars(ends_with(".padj")), ~ ifelse(is.na(.), 1, .))
vol_df.edit <- vol_df.edit <- vol_df.edit %>% mutate_at(vars(ends_with(".log2FoldChange")), ~ ifelse(is.na(.), 0, .))
```

### applies a ceiling to -log10(adp)

This may not be necessary but can help when there are a handful of genes with large -log10(adp)

```{r}
#vol_df.edit <- vol_df.edit <- vol_df.edit %>% mutate_at(vars(ends_with(".padj")), ~ ifelse(. < 1e-30, 1e-30, .))
```

### sets y axis according to max(abs(LFC))

```{r}
maxFCs <- vol_df.edit %>% summarize(across(ends_with(".log2FoldChange"), ~ max(abs(.), na.rm = TRUE)))
maxFCs
max(maxFCs)
```

### Prepare vector of plot comparisons

```{r}
comparisons_to_plot <- colnames(vol_df.edit)
comparisons_to_plot <- grep("log2FoldChange$", comparisons_to_plot, value = TRUE)
comparisons_to_plot <- sub("\\.log2FoldChange$", "", comparisons_to_plot)
comparisons_to_plot
```

### Plotting the Comparisons

```{r,fig.width=10,fig.height=10}
for (i in comparisons_to_plot) {
  # Highlight DEGs
  col_log2FoldChange <- paste0(i, ".log2FoldChange")
  col_padj <- paste0(i, ".padj")
  
  vol_df.edit <- vol_df.edit %>%
    mutate(
      toLabel = if_else(abs(vol_df.edit[[col_log2FoldChange]]) >= 1 & -log10(vol_df.edit[[col_padj]]) >= 1.3, "Gene", "Empty")
    )

  # Identify the maximum fold changefor each comparison
  maxFC <- round(max(abs(vol_df.edit[[col_log2FoldChange]])), 0)
  
  # Identify the maximum -log10 pvalue for each comparison
  
  maxP <- round(max(-log10(vol_df.edit[[col_padj]])), 0)

  sorted_df <- vol_df.edit %>%
    arrange(vol_df.edit[[col_padj]])

  # Calculate the top column
  sorted_df <- sorted_df %>%
    mutate(top = if_else(row_number() <= 25, gene_name, ""))

  # Create the volcano plot
  plot_title <- paste("Volcano Plot for", i)
  p <- ggplot(sorted_df, aes(x = -log10(.data[[col_padj]]), y = .data[[col_log2FoldChange]])) +
    geom_point(aes(color = toLabel), size = 2) +
    geom_label_repel(data = filter(sorted_df, !is.na(top)), aes(label = top), box.padding = 0.5, size = 4,max.overlaps=50) +
    scale_color_manual(values = c("grey70", "red")) +
    theme_bw() +
    geom_hline(yintercept = 0, color = "blue") +
    theme(
      panel.border = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      axis.line = element_line(colour = "black"),
      legend.position = "none"
    ) +
    scale_y_continuous(breaks = seq(-maxFC, maxFC, 1), limits = c(-maxFC-0.5, maxFC+0.5)) +
    scale_x_continuous(limits = c(0, maxP+1)) +
    ggtitle(plot_title)

  print(p)
}
```

# Upset plots 

You can also use the shiny app with the tab "Upset Plot" and paste your own lists of differentially expressed genes to produce an upset plot: 

https://yannvrb.shinyapps.io/BioMicro_data_visualization/

```{r fig.width=8,fig.height=6}
# Filter the data frame based on your initial criteria
filtered_data <- assembleDat %>% filter(MPG == "Yes" & gene_type == "protein_coding")

# Get the column names ending with .padj and .absFC
padj_columns <- names(filtered_data)[str_detect(names(filtered_data), "\\.padj$")]
absfc_columns <- names(filtered_data)[str_detect(names(filtered_data), "\\.absFC$")]

# Extract prefixes for naming the lists
prefixes <- str_remove(padj_columns, "\\.padj$")

# Create a list of data frames, each containing genes for a specific .padj column and .absFC column
gene_lists <- map2(padj_columns, prefixes, ~ {
  list_name <- .y
  fc_column <- paste0(list_name, ".absFC")
  genes <- filtered_data %>%
    filter(.data[[.x]] < 0.05 & .data[[fc_column]] > 1) %>%
    select(gene_name) %>%
    pull(gene_name)
  setNames(list(genes), list_name)
})

# Flatten the list of lists
gene_lists <- flatten(gene_lists)

# Convert the list of genes into a data frame suitable for UpSet plot
gene_lists_binary <- fromList(gene_lists)

# Generate the UpSet plot
upset(gene_lists_binary, sets = names(gene_lists), order.by = "freq")
```

# Over-representation analysis (ORA)

MIT Libraries supports our access to Metacore. Request an account here:

https://libguides.mit.edu/bioinfo/genego

## Metacore - Prepare input file

The input file for metacore is useful for other things.

This code: 
1) selects gene_name and differential expression data from the full data file
2) converts the gene_name column to rowname
3) removes all rows that contain NA in both LFC and padj columns
4) replaces any NA fold changes with 1
5) replaces any NA adp values with 0
6) round data to 3 digits
6) converts gene symbol row names to column called "GeneSym"
7) exports data to excel for upload to metacore

```{r}
metacore <- assembleDat %>%
  filter(MPG == "Yes" & gene_type == "protein_coding") %>%
  select(gene_name,matches("log2FoldChange|padj"))
metacore <- metacore %>% column_to_rownames(var = "gene_name")
metacore <- metacore %>% filter(if_any(everything(), ~ !is.na(.)))

metacore <- metacore %>% mutate_at(vars(contains("log2FoldChange")), ~ifelse(is.na(.), 0, .))
metacore <- metacore %>% mutate_at(vars(contains("padj")), ~ifelse(is.na(.), 1, .))
metacore <- round(metacore,3)
metacore <- metacore %>% tibble::rownames_to_column(var="gene_name")
write.xlsx(metacore,colNames=TRUE, quote=FALSE, rowNames=FALSE, file="220516Geh_metacore.xlsx")
```

## g:Profiler - Prepare input file

The metacore table is appropriate for use in gprofiler:
  protein coding, MPG yes genes

In addition to this R interface, a web interface for gprofiler is also available:

https://biit.cs.ut.ee/gprofiler/gost

It is interesting to compare web interface output with R results

### Setting thresholds and background set for deg selection and gprofiler

Values are set here so that adjustments to input parameters are easily made

```{r}
fc.cutoff <- 1
fdr.cutoff <- 0.05
sourceSet <- c("GO:BP", "GO:MF", "GO:CC", "KEGG", "REAC", "WP")
#sourceSet <-  c("KEGG", "REAC", "WP")
#sourceSet <-  c("KEGG")
sigFlag <- TRUE
#sigFlag <- FALSE

background <- metacore %>% dplyr::select(c("gene_name"))
```

### Write background list to a file that can be used in the web interface

```{r}
write_tsv(background, file="gpro_background.txt", col_names = FALSE)
```

### Selection of deg lists

```{r}
PRvMock.degs <- metacore %>%
  filter(run_PR_vs_Mock.padj <= fdr.cutoff & abs(run_PR_vs_Mock.log2FoldChange) >= fc.cutoff) %>% 
  dplyr::select(c("gene_name")) %>%
  mutate(degset = "degs")
```

### Write differentially expressed gene list to file for upload to gprofiler web interface

```{r}
write_tsv(as.data.frame(PRvMock.degs$gene_name), file="PRvMock.degs.txt", col_names = FALSE)
```

## Running gprofiler from R client

```{r}
PRvMock.gostres <- gost(query = PRvMock.degs$gene_name,
                              organism = "hsapiens", 
                              sources = sourceSet,significant = sigFlag, 
                              correction_method ="g_SCS", custom_bg = background$gene_name, 
                              domain_scope="custom")

PRvMock.result <- PRvMock.gostres$result %>% 
  select(c("term_name","term_id","p_value","term_size","query_size", "intersection_size"))

head(PRvMock.result %>% filter(term_size < 150) %>% arrange(desc(intersection_size)),20)
```

## Plotting gprofiler results

running with interactive = TRUE enables plotly version that responds to mouse

```{r}
gostplot(PRvMock.gostres , capped = TRUE, interactive = TRUE)
```

## Running with interactive = FALSE creates a ggplot object that can be modified and saved

```{r}
# PRvMock.gproPlot <- gostplot(PRvMock.gostres , capped = TRUE, interactive = FALSE) + 
#   ggtitle("PR v Mock")
# 
# PRvMock.gproPlot

#pdf('PRvMock_gprofilerPlot.pdf',width=8, height=5)
#gproPlot
#dev.off()
```

## Highlighting terms in plots

```{r,fig.width=8,fig.height=6}
PRvMock.selectedTerms <- head(PRvMock.result$term_id,10)
publish_gostplot(PRvMock.gproPlot, highlight_terms=PRvMock.selectedTerms,filename=NULL, width=NA, height=NA )
```

# Single sample Gene Set Enrichment Analysis (ssGSEA)

## Extract and prepare gene expression matrix in .gct format from dds object

```{r}
#Extract the normalized count matrix
normalized_counts <- as.data.frame(counts(dds, normalized = TRUE))
normalized_counts <- normalized_counts %>% rename_with(~ str_replace(., "\\.intCt$", ".normCt"), ends_with(".intCt"))

# Annotate counts table with gene information

normalized_counts <- merge(normalized_counts, gene2typesym, by.x=0, by.y="gene_id", all.x=TRUE, all.y=TRUE)

# Annotate counts table with MPG and LowExpLowVar from assembleData

expInfo <- assembleDat %>% dplyr::select(Row.names,MPG,LowExpLowVar,Avg)
normalized_counts <- merge(normalized_counts,expInfo, by.x="Row.names", by.y="Row.names", all.x=TRUE, all.y=TRUE)

# Filter to protein coding genes, MPG = yes and Avg > 0

normalized_counts <- normalized_counts %>% dplyr::filter(gene_type == "protein_coding" & MPG == "Yes" & Avg > 0)

# Reorder dataframe

normalized_counts <- normalized_counts %>% dplyr::select(gene_name,Row.names,ends_with(".normCt"))

# Rename ID columns

normalized_counts <- normalized_counts %>% dplyr::rename(NAME = gene_name, Description = Row.names)

# Prepare gct file

numDatCols <- normalized_counts %>% select(ends_with(".normCt")) %>% colnames() %>% length()
numDatRows <- normalized_counts %>% rownames() %>% length()

filename1 <- paste("d7_normCounts", ".gct", sep="")
cat("#1.2", "\n", sep="\t", file=filename1)
cat(numDatRows, numDatCols, "\n", sep="\t", file=filename1, append = TRUE)

write.table(normalized_counts, file=filename1, row.names=FALSE, quote = FALSE, sep="\t", append=TRUE)
```

## run ssGSEA

Replace by Mouse_gene_sets if needed.

```{r}
ssGSEA.project.dataset(input.ds = "d7_normCounts.gct", output.ds = "h_ssGSEA", gene.sets.dbfile.list = "Human_gene_sets/h.all.v2023.2.Hs.symbols.gmt")
ssGSEA.project.dataset(input.ds = "d7_normCounts.gct", output.ds = "c2cp_ssGSEA", gene.sets.dbfile.list = "Human_gene_sets/c2.cp.v2023.2.Hs.symbols.gmt")
ssGSEA.project.dataset(input.ds = "d7_normCounts.gct", output.ds = "c2cgp_ssGSEA", gene.sets.dbfile.list = "Human_gene_sets/c2.cgp.v2023.2.Hs.symbols.gmt")
ssGSEA.project.dataset(input.ds = "d7_normCounts.gct", output.ds = "c5bp_ssGSEA", gene.sets.dbfile.list = "Human_gene_sets/c5.go.bp.v2023.2.Hs.symbols.gmt")
ssGSEA.project.dataset(input.ds = "d7_normCounts.gct", output.ds = "c5mf_ssGSEA", gene.sets.dbfile.list = "Human_gene_sets/c5.go.mf.v2023.2.Hs.symbols.gmt")
ssGSEA.project.dataset(input.ds = "d7_normCounts.gct", output.ds = "c5cc_ssGSEA", gene.sets.dbfile.list = "Human_gene_sets/c5.go.cc.v2023.2.Hs.symbols.gmt")
```

## Import ssGSEA results, row-center the results and export to excel

Note, the Dec and WIBR2 runs are quite different so the ssGSEA data was row-centered separately in order to not mask information

```{r}
ssGSEA_file_list <- list.files(pattern = "ssGSEA.gct$", full.names = TRUE)

read_gct_file <- function(file) {
  read.table(file, header = TRUE, sep = "\t", skip = 2, stringsAsFactors = FALSE)
}

ssGSEA_gct_list <- lapply(ssGSEA_file_list, read_gct_file)
ssGSEA.results <- do.call(rbind, ssGSEA_gct_list)

# filter out "combined up and down gene sets"

ssGSEA.results <- ssGSEA.results %>% dplyr::filter(!grepl("combined UP & DN", Description))
ssGSEA.results <- ssGSEA.results %>% rename_with(~ str_replace(., "\\.normCt$", ".ssgsea"), ends_with(".normCt"))
ssGSEA.results <- ssGSEA.results %>% select(Name,ends_with(".ssgsea"))

ssGSEA.results.run1 <- ssGSEA.results %>% dplyr::select(Name,contains("run1"))
ssGSEA.results.run2 <- ssGSEA.results %>% dplyr::select(Name,contains("run2"))

ssGSEA.results.run1 <- ssGSEA.results.run1 %>% remove_rownames %>% column_to_rownames(var="Name")
ssGSEA.results.run2 <- ssGSEA.results.run2 %>% remove_rownames %>% column_to_rownames(var="Name")

ssGSEA.results.run1 <- ssGSEA.results.run1 %>% data.matrix()
rc.ssGSEA.results.run1 <- calculate_zscore(ssGSEA.results.run1)
rc.ssGSEA.results.run1.df <- as.data.frame(rc.ssGSEA.results.run1)
rc.ssGSEA.results.run1.df <- rc.ssGSEA.results.run1.df %>% rename_with(~ str_replace(., "\\.ssgsea$", ".rc_ssgsea"), ends_with(".ssgsea"))

ssGSEA.results.run2 <- ssGSEA.results.run2 %>% data.matrix()
rc.ssGSEA.results.run2 <- calculate_zscore(ssGSEA.results.run2)
rc.ssGSEA.results.run2.df <- as.data.frame(rc.ssGSEA.results.run2)
rc.ssGSEA.results.run2.df <- rc.ssGSEA.results.run2.df %>% rename_with(~ str_replace(., "\\.ssgsea$", ".rc_ssgsea"), ends_with(".ssgsea"))

rc.ssGSEA.results.df <- merge(rc.ssGSEA.results.run1.df,rc.ssGSEA.results.run2.df,by.x=0,by.y=0,all.x=TRUE,all.y=TRUE)

write.xlsx(rc.ssGSEA.results.df,colNames=TRUE, quote=FALSE, rowNames=TRUE, file="rc.ssGSEA.results.xlsx")
```

## Heatmap of specific gene sets from ssGSEA

```{r}
data_keep_rows <- c("HALLMARK_INTERFERON_ALPHA_RESPONSE","HALLMARK_INTERFERON_GAMMA_RESPONSE","KEGG_AXON_GUIDANCE")  

rc.ssGSEA.results <- rc.ssGSEA.results.df %>% remove_rownames %>% column_to_rownames(var="Row.names")

rc.ssGSEA.results.select <- rc.ssGSEA.results[rownames(rc.ssGSEA.results) %in% data_keep_rows, ]

Heatmap(rc.ssGSEA.results.select, cluster_rows = FALSE, cluster_columns = FALSE, show_row_names = TRUE)
```

# GSEA - Functional Class Sorting

## Preparing rank files for GSEA

Extract protein coding genes from assembled data files, adjust the data and write rnk file in gsea directory
this code chunk prepares rnk files for Linux GSEA runs

```{r}
system("mkdir -p ../gsea && chmod 775 ../gsea")

rnkFile <- assembleDat %>% filter(MPG == "Yes" & gene_type == "protein_coding") %>% dplyr::select('gene_name',ends_with(".stat"))

names <- colnames(rnkFile)

for (i in names[-1]){
  d<-rnkFile %>%  dplyr::select('gene_name',all_of(i))
  names(d)[1] <- '#Gene'
  d<-d[complete.cases(d), ]
  write.table(d, sep='\t',file=paste0("../gsea/",i,".rnk"),col.names=TRUE, quote=FALSE, row.names=FALSE)
}
```

## Process HPC tsv GSEA results to produce a single output file

Run bash script submit_prerank_4.3.2_v2023.2.sh to loop through each rank file for each gene set collection. Don't forget to change the paths for Mouse_gene_sets gmt files if working with mouse data.

### Identify files that need to be imported - the basedir will change according to run

```{r}
base_dir <- "../gsea/today'sdate"

# Get a list of directories in the base directory
sub_dirs <- list.dirs(base_dir, recursive = FALSE)

# Initialize an empty list to store file paths
file_list <- list()

# Loop through each sub-directory
for (sub_dir in sub_dirs) {
  # Get a list of files in the sub-directory that match the pattern
  files <- list.files(path = sub_dir, pattern = "^gsea_.*\\.tsv$", full.names = TRUE)
  
  # Add the files to the file list
  file_list <- c(file_list, files)
}
```

### Import the GSEA data into a list of lists

```{r}
gsea_results.list <- lapply(file_list, read.delim)
```

```{r}
# Define a function to process each file path
process_filepath <- function(filepath) {
  # Remove folder info from the front of the file path
  short_filename <- gsub("../gsea/today'sdate", "", filepath)
  # Replace ".rnk." with "_"
  short_filename <- gsub("\\.rnk\\.", "_", short_filename)
  # Add "_pos" if "pos" is in the remainder or "_neg" if "neg" in the remainder
  if (grepl("pos", short_filename)) {
    short_filename <- paste0(short_filename, "_pos")
  } else if (grepl("neg", short_filename)) {
    short_filename <- paste0(short_filename, "_neg")
  }
  short_filename <- gsub("\\.GseaP.*_pos$", "_pos", short_filename)
  short_filename <- gsub("\\.GseaP.*_neg$", "_neg", short_filename)
  return(short_filename)
}

# Apply the function to each file path in file_list
short_filenames <- lapply(file_list, process_filepath)
# Rename the list of gsea results
names(gsea_results.list) <- short_filenames
```

### extract relevant columns from the dataframes

```{r}
# Define the columns to select
columns_to_select <- c("NAME", "SIZE", "NES", "FDR.q.val")
# Loop through the list of dataframes
gsea_results.list <- lapply(gsea_results.list, function(df) {
  # Select the desired columns using dplyr::select
  df_selected <- dplyr::select(df, all_of(columns_to_select))
  # Return the modified dataframe
  return(df_selected)
})
```

### Add a comparison column containing the name of each results list

```{r}
# Loop through the list of dataframes
for (comparison_name in names(gsea_results.list)) {
  new_name <- gsub("_pos$|_neg$", "", comparison_name)
  # Access the dataframe by its name
  df <- gsea_results.list[[comparison_name]]
  # Add a new column containing the comparison name
  df <- df %>% mutate(Comparison = new_name)
  # Update the dataframe in the list
  gsea_results.list[[comparison_name]] <- df
}
```

```{r}
gsea_results.list <- lapply(gsea_results.list, function(x) {
  x$NES <- as.numeric(x$NES)
  return(x)
})
```

### Assemble the data into a single output file and export the results

```{r}
non_empty_dfs <- gsea_results.list[sapply(gsea_results.list, nrow) > 0]
combined_df <- bind_rows(non_empty_dfs, .id = "ComparisonFull")

#Use mutate and str_extract to create the new 'Collection' column
combined_df <- combined_df %>%
  mutate(Collection = str_extract(Comparison, "[^_]+$")) %>%
  mutate(Comparison = str_remove(Comparison, "_c2cp|_c5bp|_h|_c2cgp|_c5cc|_c5mf|_c3"))
```

You can bring the combined_df into esquisse tab of the shiny app to produce a heatmap:

https://yannvrb.shinyapps.io/BioMicro_data_visualization/

Or go ahead with pivoting script

### Pivot GSEA results

```{r}
# Get the maximum SIZE value for each NAME
df_size_max <- combined_df %>%
  group_by(NAME) %>%
  summarize(SIZE = max(SIZE, na.rm = TRUE), .groups = 'drop')

# Ensure no duplicate entries for the same NAME and Comparison
df_summary <- combined_df %>%
  group_by(NAME, Comparison, Collection) %>%
  summarize(NES = mean(NES, na.rm = TRUE), FDR.q.val = mean(FDR.q.val, na.rm = TRUE), .groups = 'drop')

# Remove suffixes like _h, _c2cp, etc. from 'Comparison' values to retain base comparison name
df_summary <- df_summary %>%
  mutate(Comparison = sub("(_h|_c[2-5].*)$", "", Comparison)) # Remove unwanted suffixes

# Pivot the dataframe to get NES and FDR.q.val with correct column names
pivoted_df <- df_summary %>%
  pivot_wider(names_from = Comparison, 
              values_from = c(NES, FDR.q.val), 
              names_sep = ".") %>%
  rename_with(~ sub("^(.+?)\\.NES$", "\\1.NES", .), starts_with("NES")) %>%
  rename_with(~ sub("^(.+?)\\.FDR.q.val$", "\\1.FDR.q.val", .), starts_with("FDR.q.val"))

# Combine the SIZE column with the pivoted dataframe
final_df <- pivoted_df %>%
  left_join(df_size_max, by = "NAME") %>%
  select(NAME, SIZE, Collection, everything()) # Ensure Collection is included and in the right order

# Identify FDR.q.val columns
fdr_columns <- grep("^FDR.q.val", names(final_df), value = TRUE)

# Create MinFDR column with the minimum value of all FDR.q.val columns
final_df <- final_df %>%
  rowwise() %>%
  mutate(MinFDR = min(c_across(all_of(fdr_columns)), na.rm = TRUE)) %>%
  ungroup()

# Sort final_df by MinFDR with smallest values at the top
final_df <- final_df %>% arrange(MinFDR)

# Clean up column names for simplicity
final_df <- final_df %>% rename_with(~ gsub("run_|\\.stat", "", .))

# Save the final dataframe
write.xlsx(final_df, "GSEA_pivot.xlsx")
```

Download the pivoted excel file, extract specific gene sets for the heatmap

```{r fig.width=4,fig.height=6}
# Remove "GOBP_" from the row names
rownames(data) <- gsub("GOBP_", "", rownames(data))

# Open the PDF device
pdf("Heatmap.pdf", width = 4, height = 6)

# Create the heatmap with reduced legend size and modified row labels
heatmap <- Heatmap(as.matrix(data),
                   name = "NES",
                   cluster_rows = TRUE,
                   cluster_columns = TRUE,
                   show_row_names = TRUE,
                   show_column_names = TRUE,
                   row_names_gp = gpar(fontsize = 3),  # Adjust font size for readability
                   column_names_gp = gpar(fontsize = 6), # Adjust font size for readability
                   heatmap_legend_param = list(
                     title_gp = gpar(fontsize = 8),  # Title font size
                     labels_gp = gpar(fontsize = 6),  # Labels font size
                     grid_width = unit(2, "mm"),      # Width of the legend color bar
                     grid_height = unit(2, "mm")      # Height of the legend color bar
                   ))

# Draw the heatmap
draw(heatmap)

# Close the PDF device
dev.off()
```

# write session info

Capturing information about the R session helps document your work

```{r}
sessionInfo()
writeLines(capture.output(sessionInfo()), "sessionInfo.txt")
```
